### PR TITLE
Remove overly restrictive check on gradient types.

### LINKED ISSUE #6858: Support for mixed precision gradients
To save memory while keeping precision it's useful to be allow functions that store activations as fp16, but compute gradients in fp32. We are working on custom ops that compute gradients in mixed mode, but it requires hacks on Python gradient type checks side to working. The feature request is to relax these checks.

Here's an example computation graph with mixed precision gradient. Forward propagation is done in fp16, and backprop is in fp32 as in the graph below 
mixed-grads

Here's a toy example that creates such a g

Graph:

```
import tensorflow as tf
from tensorflow.python.framework import function

@function.Defun(tf.float16, tf.float32)
def custom_grad(x, grad):
    return 2*tf.cast(x, tf.float32)*grad

@function.Defun(tf.float16, grad_func=custom_grad)
def custom(x):
    return tf.square(x)

x = tf.Variable(1., dtype=tf.float16)

# override cast to keep first backprop in fp32 
with tf.get_default_graph().gradient_override_map({"Cast": "Identity"}):
    loss = tf.cast(custom(x, name="mycustom_apply"), tf.float32)
    
gradient = tf.gradients(loss, x)

sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(gradient)
```

Currently the example above crashes because various places in code assumes that grad function output matches type of incoming activation.

in particular:

- [tensorflow/python/ops/gradients_impl.py](https://github.com/tensorflow/tensorflow/blob/cb17d1b0e2b581b5da1d9597b7929ba764749d38/tensorflow/python/ops/gradients_impl.py#L260), line 264, in _VerifyGeneratedGradients: enforces that gradient output type is the same as activation input type

### PR DESCRIPTION
**Pull request by yaroslavvb**
Fixes #6858 

Currently gradient computation `input.is_compatible_with(backprop)` for every input. This fails if we have float16 activations and float32 backprops (ie `tf.float32.is_compatible_with(tf.float16)` returns False), this PR removes this check.

### CODE REVIEW

**Comment by drpngx** [comment]
Should we check that we have at least numeric types?

Jenkins, test this please.

**Comment by yaroslavvb** [comment]
Makes sense, the backprops should be numeric (inputs could still be non-numeric)

**Comment by yaroslavvb** [comment]
Added another place which enforces that gradient type matches activation type. With these changes we can run scott's mixed precision backprop kernels

**Comment by girving** [comment]
It would also be nice to check that the gradient is complex iff the value is.  It might be safer to rephase the check using whitelists of similar ops to avoid missing cases like this.

**Comment by yaroslavvb** [comment]
So checking if backprops are real-valued for real-valued inputs and complex valued for complex valued inputs is more restrictive without ruling out mixed precision. I added `is_real` method to mirror `is_complex`, since the existing `is_floating` doesn't seem to cover variables or quantized types. What did you mean by whitelists of similar ops? What whitelist should I add?

**Comment by drpngx** [comment]
It's plausible to have quantized gradients for floating point parameters, I
think.

**Comment by yaroslavvb** [comment]
added checks that real-valued inputs must have real-valued backprops, allowing for quantized representation, same restriction for complex-valued.

while testing, found that some tensors (TopKV2, TensorArrayReadV3) produce integer valued gradients, so relaxed checks to allow for integer values, PTAL

**Review by girving — APPROVED** [review]

**Comment by girving** [comment]
Jenkins, test this please.

File: tensorflow/python/ops/gradients_impl.py [hunk 1]
```diff
@@ -225,6 +225,24 @@ def _DefaultGradYs(grad_ys, ys, colocate_gradients_with_ops):
         grad_ys[i] = array_ops.fill(
             array_ops.shape(y), constant_op.constant(
                 1, dtype=y.dtype))
+      continue
+    if y.dtype.is_real or y.dtype.is_integer:
+      if not grad_y.dtype.is_real and not grad_y.dtype.is_integer:
+          raise ValueError("Gradient type %s generated for real or "
```
Comment (vrv) [inline tensorflow/python/ops/gradients_impl.py]: should these be TypeErrors ?

Comment (yaroslavvb) [inline tensorflow/python/ops/gradients_impl.py]: That does sound more logical, changed

**Review by girving — APPROVED** [review]

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
Jenkins, test this please.  Maybe this time it won't abort partway through?

**Comment by gunan** [comment]
Tensorflow-sanity job is failing. Once it fails, it aborts all other jobs.

**Comment by gunan** [comment]
```
FAIL: Found 2 non-whitelited pylint errors:
tensorflow/python/ops/gradients_impl.py:232: [W0311(bad-indentation), ] Bad indentation. Found 10 spaces, expected 8

tensorflow/python/ops/gradients_impl.py:239: [W0311(bad-indentation), ] Bad indentation. Found 10 spaces, expected 8
```

**Comment by yaroslavvb** [comment]
Ooops, that's what I get for doing edits in Github edit window (bazel crashed our NFS so waiting for that to get fixed)

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Ok, so we have an issue here. We may have to revert this as is, or maybe we can patch it.

@yaroslavvb, why did you introduce a `is_real`, instead of using the `is_floating` property? More specifically, `bfloat16` is really an internal type, is it necessary to expose it?

**Comment by martinwicke** [comment]
Since there was discussion on this earlier, the bfloat16 quantization is (should be) an implementation detail for network transfer only (@josh11b, @vrv do you agree?). I don't understand the comment about Variables: in what way does is_floating not work for variables?

**Comment by vrv** [comment]
bfloat16 was indeed supposed to be a hidden optimization.  Whether that should remain that way is up to others.

**Comment by yaroslavvb** [comment]
@martinwicke I can send a PR that changes this to `is_floating`. The reason I added `is_real` is because future work may go a step further and use float32 activation, quantized backprop, which makes sense mathematically since both represent real-valued number

**Comment by martinwicke** [comment]
That's true, but let's not get ahead of ourselves, especially not with something as hacky as bfloat16.  I doubt that we'd want that type in particular for quantized backprop. A proper fix-point or short float type (half) is much more likely. 

Can you make a PR changing this to use is_floating, and remove the is_real function? Mention me so we can get this in quickly.

Thanks!
